kinesumer.go

```diff
@@ -143,22 +167,20 @@ func (k *Kinesumer) Begin() (err error) {
 		return
 	}
 
-	k.nRunning = len(shards)
+	n := k.opt.MaxShards
+	if n <= 0 || len(shards) < n {
+		n = len(shards)
+	}
+
 	k.stop = make(chan Unit, k.nRunning)
 	k.stopped = make(chan Unit, k.nRunning)
-	for _, shard := range shards {
-		worker := &ShardWorker{
-			kinesis:         k.Kinesis,
-			shard:           shard,
-			stateSync:       k.StateSync,
-			stream:          k.Stream,
-			pollTime:        k.opt.PollTime,
-			stop:            k.stop,
-			stopped:         k.stopped,
-			c:               k.records,
-			GetRecordsLimit: k.opt.GetRecordsLimit,
+	for i := 0; i < n; i++ {
+		j, err := k.LaunchShardWorker(shards)
+		if err != nil {
```

Comment:
Hmm, I think it's definitely possible not to hit the max shards. e.g. 2 processes with max_shards of 4, but there are only 7 shards. I think we should log and keep going. Not reason to stop if we failed to acquire the lock (but we should fail if we couldn't start for some other reason)